hnn_core/tests/test_extracellular.py

```diff
@@ -0,0 +1,341 @@
+                ii * len(X_p) + jj][idt] * 1e3
+            phi_p_lsa[ii][jj] = net.rec_array['grid_lsa']._data[0][
+                ii * len(X_p) + jj][idt] * 1e3
+            phi_p_theory[ii][jj] = \
+                _mathematical_dipole(e_pos, d_pos, d_Q) / conductivity
+
+    # compare the shape of the far fields
+    for phi_p in [phi_p_psa, phi_p_lsa]:
+        cosT = np.dot(phi_p.ravel(), phi_p_theory.ravel()) / (
+            norm(phi_p.ravel()) * norm(phi_p_theory.ravel()))
+        # the far field should be very close to dipolar, though threshold may
+        # need adjusting when new mechanisms are included in the cells
+        assert 1 - cosT < 1e-3
+
+    # for diagnostic plots, uncomment the following:
+    # import matplotlib.pyplot as plt
+    # from matplotlib.colors import SymLogNorm
+    # fig, axs = plt.subplots(1, 2, figsize=(8, 4))
+    # for ax, phi in zip(axs, [phi_p, phi_p_theory]):
+    #     ax.pcolormesh(X_p, Y_p, phi.T,
+    #                   norm=SymLogNorm(linthresh=1e-2, linscale=1.,
+    #                                   vmin=-5e0, vmax=5e0, base=10),
+    #                   cmap='BrBG_r', shading='auto')
+    # plt.show()
+
+
+def test_rec_array_calculation():
+    """Test LFP calculation."""
+    hnn_core_root = op.dirname(hnn_core.__file__)
+    params_fname = op.join(hnn_core_root, 'param', 'default.json')
+    params = read_params(params_fname)
+    params.update({'N_pyr_x': 3,
+                   'N_pyr_y': 3,
+                   'tstop': 25,
+                   't_evprox_1': 7,
+                   't_evdist_1': 17})
+    net = default_network(params, add_drives_from_params=True)
+
+    # one electrode inside, one above the active elements of the network
+    electrode_pos = [(1.5, 1.5, 1000), (1.5, 1.5, 3000)]
+    net.add_electrode_array('arr1', electrode_pos)
+    _ = simulate_dipole(net, n_trials=1, postproc=False)
+
+    # test accessing simulated voltages
+    assert (len(net.rec_array['arr1']) ==
+            len(net.rec_array['arr1'].voltages) == 1)  # n_trials
+    assert len(net.rec_array['arr1'].voltages[0]) == 2  # n_contacts
+    assert (len(net.rec_array['arr1'].voltages[0][0]) ==
+            len(net.rec_array['arr1'].times))
+    # ensure copy drops data (but retains electrode position information etc.)
+    net_copy = net.copy()
+    assert isinstance(net_copy.rec_array['arr1'], ExtracellularArray)
+    assert len(net_copy.rec_array['arr1'].voltages) == 0
+
+    data, times = net.rec_array['arr1'].get_data(return_times=True)
+    assert isinstance(data, np.ndarray)
+    assert isinstance(times, np.ndarray)
+    data_only = net.rec_array['arr1'].get_data()
+    assert_allclose(data, data_only)
+
+    # using the same electrode positions, but a different method: LSA
+    net.add_electrode_array('arr2', electrode_pos, method='lsa')
+
+    # make sure no sinister segfaults are triggered when running mult. trials
+    n_trials = 5  # NB 5 trials!
```

Comment:
oh man ... I really wish we don't have to do such tests.

Reply:
The problem with slow tests like this and the [test above](https://github.com/jonescompneurolab/hnn-core/pull/329/files#r653085282) is that it will discourage people from running them locally. Ideally, you want to keep your test suite run in under a minute

Reply:
You mean the _entire_ test suite? We may have a problem with `test_network` and `test_parallel_backends` then...

The extracellular tests run in 20-25 s on my Air, which I can deal with. The good thing about extracellular is that it shouldn't interact with the rest of the code, so it can be run in isolation while developing. But if the goal is more like 5 s, I need to tighten things up...

Reply:
Btw. `n_trials = 2` wasn't enough to catch my troubles with `extra_scatter_gather`!

Reply:
Ideally, the entire test suite yes :) Of course it's the ideal that we won't achieve. When I'm making some tricky changes, I run the entire tests after every change to verify things don't break. You guys got to start running more tests locally rather than relying on CIs ;-)